File: min_element_RotatedArray.py
# Time Complexity : Add - O(log n)
# Space Complexity :O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

'''
I used modified binary search to find a pivot element, element such that A[i] > A[i+1]

The code also works for a sorted array(just to account for 180 degree rotation), by returning first element if that case
occurs

'''

class Solution:
    
    def pivot_element(self, nums):
        
        low = 0
        high = len(nums) - 1
        
        while (low <= high):
            mid = low + int((high - low)//2)
            if mid < high and nums[mid] > nums[mid+1]:
                return nums[mid+1]      
            elif mid > low and nums[mid-1] > nums[mid]:
                return nums[mid]         
            elif nums[low] <= nums[mid]:
                low = mid + 1
            else:
                high = mid - 1                
        else:
            return nums[0]
        
        
    def findMin(self, nums):
        
        if nums == [] and len(nums) < 1:
            return
            
        pivot_element = self.pivot_element(nums)
        return pivot_element
        
        
                

File: min_element_rotated_array.java
// time: O(log n)
//space: O(1)

class min_element_rotated_array {
    public int findMin(int[] nums) {
        int minEle = -1;
        
        int start = 0;
        int end = nums.length - 1;
        
        if(nums.length == 1){
            return nums[0];
        }
        
        if(nums[end] > nums[start]){
            return nums[start];
        }
        
        while(start <= end){
            int mid = start + (end -start)/2;
            // mid element is greater than the next element, next is the minEle
            if(nums[mid] > nums[mid +1]){
                return nums[mid + 1];
            } 
            
// mid ele is less than the previous element, mid is the minnEle
            if(nums[mid] < nums[mid-1]){
                return nums[mid];
            }
            
            if(nums[mid] > nums[start]){
                start = mid +1;
            } else {
                end = mid - 1;
            }
        }
        return -1;
    }
}